src/dsl/compute_graph.jl

```diff
@@ -0,0 +1,180 @@
+using AbstractTrees
+
+# In this file we define a tree data structure which can be used for optimisation passes
+# over contraction commands
+
+"""Generic binary tree node data structure"""
+mutable struct BinaryNode{T}
+    data::T
+    parent::BinaryNode{T}
+    left::BinaryNode{T}
+    right::BinaryNode{T}
+
+    # Root constructor
+    BinaryNode{T}(data) where T = new{T}(data)
+    # Child node constructor
+    BinaryNode{T}(data, parent::BinaryNode{T}) where T = new{T}(data, parent)
+end
+BinaryNode(data) = BinaryNode{typeof(data)}(data)
+
+"""
+    leftchild(data, parent::BinaryNode)
+
+Returns left child node or nothing
+"""
+function leftchild(data, parent::BinaryNode)
```

Comment:
Can this signature be changed to `leftchild!(parent, data)` as it modifies the parent node, or is it adding a function method to an interface function in AbstractTrees?

Reply:
Good point! These functions are taken from the example provided with AbstractTrees at `https://github.com/JuliaCollections/AbstractTrees.jl/blob/master/examples/binarytree_core.jl`. I assumed these were just accessing the variables but now that I look at them I think they were intended to update the items. These are not used currently so can safely remove them.